 import { CHANGE_FOCUS, CHANGE_PAGE, SLIDING_END } from "../constants/action-types";
import Image1 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Title/titled.jpg';
import Image2 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-01.jpg';
import Image3 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-02.jpg';
import Image4 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-02-02.png';
import Image5 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TM100-popup-1.jpg';
import Image6 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-003.jpg';
import Image7 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-003_02.png';
import Image8 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-03.jpg';
import Image9 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-04.png';
import Image10 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-05.jpg';
import Image11 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-06.png';
import Image12 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-07.jpg';
import Image13 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-08.png';
import Image14 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-09.jpg';
import Image15 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-10.jpg';
import Image16 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-11.jpg';
import Image17 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP100.jpg';
import Image18 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP101.jpg';
import Image19 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP102.jpg';
import Image20 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-12.jpg';
import Image21 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-13.jpg';
import Image22 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP103.jpg';
import Image23 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-14.jpg';
import Image24 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/file-01.jpg';
import Image25 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-15.jpg';
import Image26 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP105.jpg';
import Image27 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-16.jpg';
import Image28 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-17.jpg';
import Image29 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-18.jpg';
import Image30 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-19.jpg';
import Image31 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-20.jpg';
import Image32 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-20-02.png';
import Image33 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-21.jpg';
import Image34 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-21-02.png';
import Image35 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP106.jpg';
import Image36 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-22.jpg';
import Image37 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-22-02.jpg';
import Image38 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-23.jpg';
import Image39 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-24.jpg';
import Image40 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-25.jpg';
import Image41 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-26.jpg';
import Image42 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-27.jpg';
import Image43 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-28.jpg';

import Image44 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-01.jpg';
import Image45 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-02-01.jpg';
import Image46 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-03-01.jpg';
import Image47 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-04-01-01.jpg';
import Image48 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-05-01-01-01.jpg';
import Image49 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-06-01-01-01-01.jpg';
import Image50 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-07-01-01-01-01-01-01.jpg';

import Image51 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TMMMM-01.jpg';


const initialState = {
  pageData:[
    
 {
      'title': 'Module 1 Sections Menu:',
      'description': 'Understanding the Basic Visual Language of TM - using a hypothetical TM \n Viral genome \n Terminal repeats \n Visual scale \n RNA transcript \n RNA/DNA directionalities \n Color coding & Text labels \n \n Visual Summary \n Quiz- module 1',
      
      'image': Image1
    },

{
      'title': 'What does a TM look like?',
      'description': 'This figure is a typical TM example that summarized all the transcription information about a DNA virus called the adenovirus. You can see this TM contains many different graphical features and they are difficult to understand at the first time. I will not teach you the visual language by going through this graph. Instead, let’s start with a simple hypothetical example of TM.',
      'image': Image2
    },


  {
      'title': 'Using a hypothetical TM to understanding the Basic Visual Language - the viral genome',
      'description': 'When you first look at the TM, you should read from the fundamental element that every TM has: the viral genome. In this case, the two black lines represent the viral genome. By moving the slider, you can see the double-stranded helical viral DNA is simplified into two black lines in TM, which reduces the visual complexity of the DNA double helix as well as saves the design space.',
      
      'image': Image3,
      'afterImage': Image4
    },


{
      
      'description': 'other viruses may have linear genome, cyclized, or partially double-strand DNA genome, or RNA genome etc, and they can be presented by using lines in slightly different ways.',
      'image': Image5
    },


 {
      'title': 'Using a hypothetical TM to understanding the Basic Visual Language - terminal repeats',
      'description': 'You may noticed that both ends of the viral genome can have these black boxes highlighted in purple (move the slider). These are usually represent the terminal repeats (TR).',
      
      'image': Image6,
      'afterImage': Image7
    },


{
      
      'description': 'Usually in viral genome, we can either have the inverted terminal repeats (ITR) or the long direct/flanking terminal repeat (LTRs). The schematic diagram illustrates the ITR in viral genome：the ITR sequences at two ends of the same strand are complementary to each other, and the ITR in one strand is bind to its reverse complementary repeats in other strand.',
      'image': Image8
    },

{
    
      'description': 'To test this, I want you to separate this double stranded genome apart.',
      'image': Image9
    },


{
      
      'description': 'After the separation, try to fold each DNA strand as the figure shown below and to observe the result.',
      'image': Image10
    },


{
     
      'description': 'As you can see, the terminal inverted repeats are capable of self-complementary base pairing. The biological function of this self base pairing is to reform a replication origin as we observed in double strand DNA. Viruses can either use single strand or double strand DNA genome to improve the genome replication efficiency.',
      'image': Image11
    },

{
      
      'description': 'But long direct terminal repeats (LTRs) have different repeat sequence.',
      'image': Image12
    },


{
     
      'description': 'Then, I want you to separate this double stranded genome as well.',
      'image': Image13
    },

{
      
      'description': 'After separation, you may realize that they cannot form base pairs when you try to fold each strand like this. This is one way to distinguish LTR from ITR. Both of these two repeats have different biological functions. Click the external HTML link to explore more about these two repeats.',
      'image': Image14
    },


    

{
      'title': 'Using a hypothetical TM to understanding the Basic Visual Language - Visual scale',
      'description': 'Then, a numerical visual scale is present beside to the viral genome, indicating the total length of the genome, and providing the accurate coordinates of all RNA transcripts relative to their corresponding viral genome parts. This visual scale is similar to a ruler that is usually marked off in specific units. There are different styles for the measurement of viral genome.',
    
      'image': Image15,
      'popup': {
        'title': 'different visual styles',
        'content': [
          {
            'description': 'The scale can be a ruler only or with labels (unit) adjacent to the genome. In some cases, a table depicting the regions of interest may be given to show the precise positions of RNA transcripts.',
            'image': Image16
          },
          {
            'description': 'Sometimes the scale shows the nucleotide number (nt) to measure the single stranded or partially single stranded viral genome because there is no base pair in single stranded region.  ',
            'image': Image17
          },
          {
            'description': 'The unit of the scale can also be kilo-base pair (kb).',
            'image': Image18
          },
          {
            'description': 'In a different scenario, these numbers are normalized to 100 map unit. For example, adenovirus genome size equals to 100 map units, and each map unit equals to 350 nucleotides. Therefore, the total length of adenovirus genome is 35000 nucleotides, or 35000 bp (350 bp * 100 map units). ',
            'image': Image19
          }
          
        ]
      }
    },


{
      'title': 'Using a hypothetical TM to understanding the Basic Visual Language - RNA transcript',
      'description': 'Adjacent to double-stranded DNA genome, you will find an orange arrow represents the RNA transcript transcribed from this viral genome. As you know, a typical arrow contains an arrowhead, the body of the arrow, an arrow tail.',
      'image': Image20
    },



{
    
      'description': 'Before we discuss about the arrowhead and tail, I want you to understand that RNA has a specific directionality: 5’ and 3’ ends.',
    
      'image': Image21,
      'popup': {
        'title': 'DNA & RNA directionalities',
        'content': [
          {
            'description': 'We name the directionality according to the sugar backbone of nucleotides either in RNA or DNA. Since they have exactly the same chemical structure at the third and fifth prime carbon of the ribose ring, the DNA molecule is used here to explain the directionality.The inset of this illustration explains the locations of 5’ and 3’ carbons in the deoxyribose. These two carbons of sugar molecules can connected to each other using phosphate to form DNA backbone in the middle of the strand. However, at both ends of the strands, the 5′-end contains a free phosphate group attached to the 5′ carbon of the ribose ring whereas the 3′-end has an free -OH group attached to the 3’ carbon.',
            'image': Image22
          }
          
        ]
      }
    },




{
    
      'description': 'Remember this directionality of RNA and let’s start with the arrowhead of RNA.The Arrowhead usually has the dual graphical functions: 1.The direction of transcription: from 5’ to 3’. 2.3’ end of mRNA transcript, corresponding to the polyadenylation site, which is a type of transcriptional modification created by the host cell mechanism.',
    
      'image': Image23,
      'popup': {
        'title': '3 prime poly-A tail',
        'content': [
          {
            'description': 'As you can see, after transcription, a precursor RNA is being made, and it contains two essential signal sequences for polyadenylation. The signal sequences include an AAUAAA poly-A addition site sequence and a GU-rich sequence located at downstream. These signals will recruit specific enzymes to make a cut between these two sequences, and create an novel OH 3’ end 10-35 nts downstream from the AAUAAA sequence. Cleaved GU-rich sequence is then degraded. This newly made free OH end provides a poly-A addition site for poly-A polymerase.',
            'image': Image24
          }
          
        ]
      }
    },

    {
    
      'description': 'Because RNA transcript synthesis is from 5’ to 3’. We can deduce that the arrow tail represents the 5’ end of the transcript. To visualize this, TMs usually use a vertical bar to represent the 5’ end of RNA, corresponding to transcriptional promoter/initiation site of transcription as well as another type of transcriptional modification: 5’ capping.',
    
      'image': Image25,
      'popup': {
        'title': '5 prime cap',
        'content': [
          {
            'description': 'The great majority of viral and cellular mRNAs possess the same 5’ cap: m7G-ppp- N (any nucleotide). m7G stands for a methyl group attaches to the 7th nitrogen of guanosine. ppp is 5’,5’-triphosphate bridge. This bridge connects the m7G with any nucleotide (N) of the 5’ end of mRNA primary transcript.',
            'image': Image26
          }
          
        ]
      }
    },

{
      
      'description': 'You also can see the body of RNA transcript shown as a straight line.',
      'image': Image27
    },

{
      
      'description': 'However, in a typical TM, you will only see the RNA transcripts in a discontinuous form: the line with interruptions/broken lines, which means the processed RNA transcripts. So what does these gaps mean?',
      'image': Image28
    },

    {
      
      'description': 'Let’s check the broken orange lines first; they are exons of the processed RNA.',
      'image': Image29
    },

{
      
      'description': 'The gaps are highlighted in light blue, and they are the removed introns.',
      'image': Image30
    },


{
    
      'description': 'Typically, in TMs, the process of intron removal and transcriptional modifications are not explained. Thus, the detailed mechanism will be illustrated here (move the slider): Step 1. The 5’ cap is synthesized to protect RNA from exonucleolytic attack as well as to facilitate transcription and translation. Remaining part of the precursor RNA is transcribed from the viral genome in 5’ to 3’ direction.',
      
      'image': Image31,
      'afterImage': Image32
    },

{
    
      'description': 'Step 2. introns are removed/ not explicitly shown in TM',
    
      'image': Image33,
      'afterImage': Image34,
      'popup': {
        'title': 'Splicing visual representation',
        'content': [
          {
            'description': 'Other visual methods indicated the removed introns include dotted line, and interruptions that are connected by a “V” shape.',
            'image': Image35
          }
          
        ]
      }
    },

{
    
      'description': 'Step 3. It’s important to note that in vivo, the exons will be joined back together, and produce a shorter RNA transcript with a poly-A tail whereas this process is also not visually shown in most of the TMs.',
      
      'image': Image36,
      'afterImage': Image37
    },

{
      
      'description': 'Therefore, you may find out the RNAs in TM are usually illustrated as the intermediate step of RNA processing as shown in the purple box.',
      'image': Image38
    },

{
      
      'description': 'Why only showing the intermediate step of RNA processing in TM? The major advantage of this step is to help audiences to locate the exact positions of exons and removed introns to the viral genome. In this figure, you can use the guide lines to determine the length and position of every exons as well as introns.',
      'image': Image39
    },

{
      
      'description': 'Now, let’s refresh your memory and improve your understanding: please check all 5’ ends in this figure.',
      'image': Image40
    },

{
      
      'description': 'A common misconception would be defining the 5 prime end of RNA transcripts in TM. If you see the 5 prime end of RNA without a vertical bar or any visual cue, do not be confused, this is not the actual 5 prime end or transcription initiation site, instead, the interruption between RNA transcript, corresponding to the removed intron.',
      'image': Image41
    },

{
      'title': 'Using a hypothetical TM to understanding the Basic Visual Language - Determine RNA/DNA directionalities',
      'description': 'We already know the directionality of an RNA transcript, but what about the directionality of each DNA strand of the viral genome?',
      'image': Image42
    },

{
    
      'description': 'You can determine the directions of DNA double strands by following this two-step process. Steps: 1. Look at your RNA and find its 3’ end (arrow head) and 5’ end (arrow tail). 2. Find the DNA single strand that adjacent to this RNA strand and this DNA strand reads in the opposite direction. Its complementary DNA strand has the same direction as the RNA strand.',
    
      'image': Image43,
      'popup': {
        'title': 'Viral genome directionalities',
        'content': [
          {
            'description': 'These two DNA strands have different names during the transcription. In this case shown in previous slide: the top strand is referred to the coding strand (anti-sense), which serves as a template during transcription. The bottom strand is called the complementary strand (sense), which sequence is same as the RNA sequence (although all the Ts in RNA are replaced by Us).',
            'image': Image44
          },
           {
            'description': 'Coding strand is the only strand used in transcription since RNA polymerase can’t use both strands at each round of gene expression.',
            'image': Image45
          },
           {
            'description': 'Step1: When the transcription begins, the RNA polymerase will bind to DNA double helix.',
            'image': Image46
          },
           {
            'description': 'Step2: RNA polymerase will then unwind part of the DNA to find the coding strand.',
            'image': Image47
          },
           {
            'description': 'Step3: RNA polymerase will separate part of dsDNA to form a bubble like structure indicated as strands separation. let’s say you want to transcribe the gene from the bubble part of the DNA.',
            'image': Image48
          },
           {
            'description': 'Step4: RNA polymerase will scan through the coding DNA strand (antisense DNA) from 3’ to 5’, and RNA synthesis is from 5’ to 3’.',
            'image': Image49
          },
           {
            'description': 'Step5: Since newly synthesized RNA is complementary to the coding strand, scientists can hybridize the RNAs with their coding DNA strands to deduce the genome positions of these RNAs  – this experiment is also known as DNA-RNA hybridization. This is why the TM was designed in a way that all RNA transcripts are adjacent to their coding strands.',
            'image': Image50
          },
          
        ]
      }
    },

{
      
      'description': 'The directionality and the name of the DNA strands are relative to the directionality of the RNA transcript since the transcription can occurs at the both DNA strands, but in opposite direction. Please check the figure shown on your left side, The complementary strand for the top RNA transcript would be the coding strand for the bottom RNA transcript.',
      'image': Image51
    }
  ],
  focus: 0,
  currentPage: 0,
  sliding: false
};

function rootReducer(state = initialState, action) {
  if (action.type === CHANGE_FOCUS) {
    if (action.focus >= 0 && action.focus < state.pageData.length) {
      state = Object.assign({}, state, {
          focus: action.focus
      });
    }
  }
  else if (action.type === CHANGE_PAGE) {
    if (action.currentPage >= 0 && 
        action.currentPage < state.pageData.length) {
      state = Object.assign({}, state, {
          currentPage: action.currentPage,
          sliding: true
      });
    }
  }
  else if (action.type === SLIDING_END) {
    clearInterval(action.intervalID);
    state = Object.assign({}, state, {
        sliding: false
    });
  }

  return state;
}

export default rootReducer;