 import { CHANGE_FOCUS, CHANGE_PAGE, SLIDING_END } from "../constants/action-types";

 //module1
import Image1 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Title/titled.jpg';
import Image2 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-01.jpg';
import Image3 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-02.jpg';
import Image4 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-02-02.png';
import Image5 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TM100-popup-1.jpg';
import Image6 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-003.jpg';
import Image7 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-003_02.png';
import Image8 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-03.jpg';
import Image9 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-04.png';
import Image10 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-05.jpg';
import Image11 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-06.png';
import Image12 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-07.jpg';
import Image13 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-08.png';
import Image14 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-09.jpg';
import Image15 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-10.jpg';
import Image16 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-11.jpg';
import Image17 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP100.jpg';
import Image18 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP101.jpg';
import Image19 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP102.jpg';
import Image20 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-12.jpg';
import Image21 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-13.jpg';
import Image22 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP103.jpg';
import Image23 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-14.jpg';
import Image24 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/file-01.jpg';
import Image25 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-15.jpg';
import Image26 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP105.jpg';
import Image27 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-16.jpg';
import Image28 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-17.jpg';
import Image29 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-18.jpg';
import Image30 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-19.jpg';
import Image31 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-20.jpg';
import Image32 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-20-02.png';
import Image33 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-21.jpg';
import Image34 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-21-02.png';
import Image35 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/pop-up_windows/TMPOP106.jpg';
import Image36 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-22.jpg';
import Image37 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-22-02.jpg';
import Image38 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-23.jpg';
import Image39 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-24.jpg';
import Image40 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-25.jpg';
import Image41 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-26.jpg';
import Image42 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-27.jpg';
import Image43 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-28.jpg';
import Image52 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TM-29-01-01.jpg';


import Image44 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-01.jpg';
import Image45 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-02-01.jpg';
import Image46 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-03-01.jpg';
import Image47 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-04-01-01.jpg';
import Image48 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-05-01-01-01.jpg';
import Image49 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-06-01-01-01-01.jpg';
import Image50 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/Pop-up_Strand_definition/DNAintro-07-01-01-01-01-01-01.jpg';

import Image51 from '../../assets/Viraliterate_Module 1_Thumbnails_V01/TM_intro_and_terminal_repeats/Main_interface/TMMMM-01.jpg';

//Module 0
import Image53 from '../../assets/Module 0_V01/Module01-01.jpg';
import Image54 from '../../assets/Module 0_V01/Module02-01-01.jpg';

import Image55 from '../../assets/Module 0_V01/Module03-01.jpg';
import Image56 from '../../assets/Module 0_V01/Module04-01.jpg';
import Image57 from '../../assets/Module 0_V01/Module05-01.jpg';
import Image58 from '../../assets/Module 0_V01/Module06-01.jpg';
import Image59 from '../../assets/Module 0_V01/Module07-01.jpg';
import Image60 from '../../assets/Module 0_V01/Module08-01.jpg';
import Image61 from '../../assets/Module 0_V01/Module09-01.jpg';

//module 2
import Image62 from '../../assets/Module2_V01/Apply_module2_ViraLiterate/Apply_module2_ViraLiterate_revised_V002.001.jpeg';
import Image63 from '../../assets/Module2_V01/Apply_module2_ViraLiterate/Apply_module2_ViraLiterate_revised_V002.002.jpeg';
import Image64 from '../../assets/Module2_V01/images/Module2-01-01.jpg';
import Image65 from '../../assets/Module2_V01/images/Module2-02-01.jpg';
import Image66 from '../../assets/Module2_V01/images/Module2-03-01.jpg';
import Image67 from '../../assets/Module2_V01/images/Module2-04-01.jpg';
import Image68 from '../../assets/Module2_V01/images/Module2-05-01.jpg';
import Image69 from '../../assets/Module2_V01/images/Module2-06-01.jpg';
import Image70 from '../../assets/Module2_V01/images/Module2-07-01.jpg';
import Image71 from '../../assets/Module2_V01/images/Module2-08-01.jpg';
import Image72 from '../../assets/Module2_V01/images/Ai file-01.jpg';
import Image73 from '../../assets/Module2_V01/images/Module2-10-01.jpg';
import Image74 from '../../assets/Module2_V01/images/Module2-09-01.jpg';
import Image75 from '../../assets/Module2_V01/images/Module2-11.jpg';
import Image76 from '../../assets/Module2_V01/images/Module2-12.jpg';

import Image77 from '../../assets/Module2_V01/images/Module2-13-01.jpg';
import Image78 from '../../assets/Module2_V01/images/Module2-14-01.jpg';
import Image79 from '../../assets/Module2_V01/images/Module2-15-01.jpg';
import Image80 from '../../assets/Module2_V01/images/Module2-16-01.jpg';
import Image81 from '../../assets/Module2_V01/images/Module2-17-01.jpg';
import Image82 from '../../assets/Module2_V01/images/Module2-18-01.jpg';
import Image83 from '../../assets/Module2_V01/images/Module2-19-01.jpg';

import Image84 from '../../assets/Module2_V01/images/Module2-20-01.jpg';
import Image85 from '../../assets/Module2_V01/images/Module2-22-01.jpg';
import Image86 from '../../assets/Module2_V01/images/Module2-23-01.jpg';
import Image87 from '../../assets/Module2_V01/images/Module2-24-01.jpg';
import Image88 from '../../assets/Module2_V01/images/Apply_module2_V002.030.jpeg';
import Image89 from '../../assets/Module2_V01/images/Module2-25-01.jpg';
import Image90 from '../../assets/Module2_V01/images/Module2-37-01.jpg';
import Image91 from '../../assets/Module2_V01/images/Module2-36-01.jpg';
import Image92 from '../../assets/Module2_V01/images/Module2-26-01.jpg';
import Image93 from '../../assets/Module2_V01/images/Module2-27-01.jpg';
import Image94 from '../../assets/Module2_V01/Apply_module2_ViraLiterate/Apply_module2_ViraLiterate_revised_V002.036.jpeg';
import Image95 from '../../assets/Module2_V01/Apply_module2_ViraLiterate/Apply_module2_ViraLiterate_revised_V002.037.jpeg';

import Image96 from '../../assets/Module2_V01/images/Module2-28-01.jpg';
import Image97 from '../../assets/Module2_V01/images/Module2-35-01.jpg';
import Image98 from '../../assets/Module2_V01/images/Module2-34-01.jpg';

import Image99 from '../../assets/Module2_V01/images/Module2-32-01.jpg';
import Image100 from '../../assets/Module2_V01/images/Module2-33-01.jpg';
import Image101 from '../../assets/Module2_V01/images/Module2-30-01.jpg';
import Image102 from '../../assets/Module2_V01/images/Module2-29-01.jpg';
import Image103 from '../../assets/Module2_V01/images/Module2-31-01.jpg';

import animationData1 from '../../assets/Animation/animation1.json';

const initialState = {
  pageData:[
    {
      animation: animationData1
    },
    {
      animation: animationData1
    },
//module0 
 {
     
      'title': 'Module 0',
      'description': 'Introduction module, page-01',
      'image': Image53
    },

    {
     
      'title': 'Module 0',
      'description': 'Introduction module, page-02',
      'image': Image54
    },

{
     
      'title': 'Module 0',
      'description': 'Introduction module, page-03',
      'image': Image55
    },
    {
     
      'title': 'Module 0',
      'description': 'Introduction module, page-04',
      'image': Image56
    },
    {
     
      'title': 'Module 0',
      'description': 'Introduction module, page-05',
      'image': Image57
    },
    {
     
      'title': 'Module 0',
      'description': 'Introduction module, page-06',
      'image': Image58
    },
{
     
      'title': 'Module 0',
      'description': 'Introduction module, page-07',
      'image': Image59
    },
{
     
      'title': 'Module 0',
      'description': 'Introduction module, page-08',
      'image': Image60
    },
{
     
      'title': 'Module 0',
      'description': 'Introduction module, page-09',
      'image': Image61
    },

  


  //module 1
    
 {
      'title': 'Module 1 Sections Menu:',
      'description': 'What does a TM look like? \n \n Understanding the Basic Visual Language of TM - using a hypothetical TM: \n \n 1. Viral genome \n 2. Terminal repeats \n 3. Visual scale \n 4. RNA transcript \n 5. RNA/DNA directionalities \n \n 6. Visual Summary \n 7. Quiz- module 1',
      
      'image': Image1
    },

{
      'title': 'What does a TM look like?',
      'description': 'This figure is a typical TM example that summarized all the transcription information about a DNA virus called the adenovirus. You can see this TM contains many different graphical features and they are difficult to understand at the first time. \n I will not teach you the visual language by going through this graph. Instead, let’s start with a simple hypothetical example of TM.',
      'image': Image2
    },


  {
      'title': 'Using a hypothetical TM to understanding the Basic Visual Language - the viral genome',
      'description': 'When you first look at the TM, you should read from the fundamental element that every TM has: the viral genome. \n In this case, the two black lines represent the viral genome. By moving the slider, you can see the double-stranded helical viral DNA is simplified into two black lines in TM, which reduces the visual complexity of the DNA double helix as well as saves the design space.',
      
      'image': Image3,
      'afterImage': Image4
    },


{
      
      'description': 'Other viruses may have linear genome, cyclized, or partially double-strand DNA or RNA genome etc, and they can be presented by using lines in slightly different ways. \n \n Click the external links in the pop up window to explore more. ',
      'image': Image5,
     'popup': {
        'title': 'Different types of viral genome',
        'content': [
          {
            'description': 'We only provide three examples here and you can also find other viruses in ths website - ViralZone: \n \n Parvoviridae (single stranded viral genome):              https://viralzone.expasy.org/103?outline=all_by_species \n \n Polyomaviridae (cyclized double stranded viral genome):              https://viralzone.expasy.org/148 \n \nHepadnaviridae (partially double stranded viral genome):              https://viralzone.expasy.org/9 \n \n (still need web development here - hyperlink function)',
            
          }
          
        ]
      }

    },




{
    'title': 'Using a hypothetical TM to understanding the Basic Visual Language - terminal repeats',
      'description': 'You may noticed that both ends of the viral genome can have these black boxes highlighted in purple (move the slider). These are usually represent the terminal repeats (TR).',

    
      'image': Image6,
      'afterImage': Image7,
      'popup': {
        'title': 'Terminal repeats',
        'content': [
          {
            'description': 'Usually in viral genome, we can either have the inverted terminal repeats (ITR) or the long direct/flanking terminal repeat (LTRs). \n The schematic diagram illustrates the ITR in viral genome：the ITR sequences at two ends of the same strand are complementary to each other, and the ITR in one strand binds to its reverse complementary repeats in other strand.',
      'image': Image8
          },

           {
            'description': 'To test this, I want you to separate this double stranded genome apart.',
      'image': Image9
          },
           {
            'description': 'After the separation, try to fold each DNA strand as the figure shown below and to observe the result.',
      'image': Image10
          },
           {
            'description': 'As you can see, the terminal inverted repeats are capable of self-complementary base pairing. The biological function of this self base pairing is to reform a replication origin as we observed in double strand DNA. Viruses can either use single strand or double strand DNA genome to improve the genome replication efficiency.',
      'image': Image11
          },
           {
            'description': 'But long direct terminal repeats (LTRs) have different repeat sequence.',
      'image': Image12
          },
           {
            'description': 'Then, I want you to separate this double stranded genome as well.',
      'image': Image13
          },
           {
            'description': 'After separation, you may realize that they cannot form base pairs when you try to fold each strand like this. This is one way to distinguish LTR from ITR. \n Both of these two repeats have different biological functions. Click the external HTML link to explore more about these two repeats (still need web development here).',
      'image': Image14
          },

        ]
      }
    },


{
      'title': 'Using a hypothetical TM to understanding the Basic Visual Language - Visual scale',
      'description': 'Then, a numerical visual scale is presented beside to the viral genome, indicating the total length of the genome. The scale provides the accurate coordinates of all RNA transcripts relative to their corresponding viral genome parts. This visual scale is similar to a ruler that is usually marked off in specific units. There are different styles for the measurement of viral genome.',
    
      'image': Image15,
      'popup': {
        'title': 'different visual styles',
        'content': [
          {
            'description': 'The scale can be a ruler only (1) or with labels (unit) adjacent to the genome (2). In some cases(3), a table depicting the regions of interest may be given to show the precise positions of RNA transcripts.',
            'image': Image16
          },
          {
            'description': '\n \n Sometimes the unit of the scale shows the nucleotide number (nt) to measure the single stranded or partially single stranded viral genome because there is no base pair in single stranded region.  ',
            'image': Image17
          },
          {
            'description': 'The unit of the scale can also be kilo-base pair (kb).',
            'image': Image18
          },
          {
            'description': 'In a different scenario, these numbers are normalized to 100 map unit. For example, adenovirus genome size equals to 100 map units, and each map unit equals to 350 nucleotides. Therefore, the total length of adenovirus genome is 35000 nucleotides, or 35000 bp (350 bp * 100 map units). ',
            'image': Image19
          }
          
        ]
      }
    },


{
      'title': 'Using a hypothetical TM to understanding the Basic Visual Language - RNA transcript',
      'description': 'Adjacent to double-stranded DNA genome, you will find an orange arrow represents the RNA transcript transcribed from this viral genome. \n As you know, a typical arrow contains an arrowhead, the body of the arrow, an arrow tail.',
      'image': Image20
    },



{
    
      'description': 'Before we discuss about the arrowhead and tail, I want you to understand that RNA has a specific directionality: 5’ and 3’ ends.',
    
      'image': Image21,
      'popup': {
        'title': 'DNA & RNA directionalities',
        'content': [
          {
            'description': 'We name the directionality according to the sugar backbone of nucleotides either in RNA or DNA. Since DNA and RNA have exactly the same chemical structure at the third and fifth prime carbon of the ribose ring, the DNA molecule is used here to explain the directionality.The inset of this illustration explains the locations of 5’ and 3’ carbons in the deoxyribose. These two carbons of sugar molecules can connected to each other using phosphate to form DNA backbone in the middle of the strand. However, at both ends of the strands, the 5′-end contains a free phosphate group attached to the 5′ carbon of the ribose ring whereas the 3′-end has an free -OH group attached to the 3’ carbon.',
            'image': Image22
          }
          
        ]
      }
    },




{
    
      'description': 'Remember this directionality of RNA and let’s start with the arrowhead of RNA.The Arrowhead usually has the dual graphical functions:\n  1. The direction of transcription: from 5’ to 3’. \n 2. 3’ end of mRNA transcript, corresponding to the polyadenylation site, which is a type of transcriptional modification created by the host cell mechanism.',
    
      'image': Image23,
      'popup': {
        'title': '3 prime poly-A tail',
        'content': [
          {
            'description': 'As you can see, after transcription, a precursor RNA is being made, and it contains two essential signal sequences for polyadenylation. The signal sequences include an AAUAAA poly-A addition site sequence and a GU-rich sequence located at downstream. These signals will recruit specific enzymes to make a cut between these two sequences, and create an novel OH 3’ end 10-35 nts downstream from the AAUAAA sequence. Cleaved GU-rich sequence is then degraded. This newly made free OH end provides a poly-A addition site for poly-A polymerase.',
            'image': Image24
          }
          
        ]
      }
    },

    {
    
      'description': 'Because RNA transcript synthesis is from 5’ to 3’. We can deduce that the arrow tail represents the 5’ end of the transcript. To visualize this, TMs usually use a vertical bar to represent the 5’ end of RNA, corresponding to transcriptional promoter/initiation site of transcription as well as another type of transcriptional modification: 5’ capping.',
    
      'image': Image25,
      'popup': {
        'title': '5 prime cap',
        'content': [
          {
            'description': 'The great majority of viral and cellular mRNAs possess the same 5’ cap: m7G-ppp- N (any nucleotide). m7G stands for a methyl group attaches to the 7th nitrogen of guanosine. ppp is 5’,5’-triphosphate bridge. This bridge connects the m7G with any nucleotide (N) of the 5’ end of mRNA primary transcript.',
            'image': Image26
          }
          
        ]
      }
    },

{
      
      'description': 'You also can see the body of RNA transcript shown as a straight line.',
      'image': Image27
    },

{
      
      'description': 'However, in a typical TM, you will only see the processed RNA transcripts in broken lines. So what do these gaps/interruptions mean?',
      'image': Image28
    },

    {
      
      'description': 'Let’s check the broken orange lines first; they are exons of the processed RNA.',
      'image': Image29
    },

{
      
      'description': 'The gaps are highlighted in light blue, and they are the removed introns.',
      'image': Image30
    },


{
    
      'description': 'Typically, in TMs, the process of intron removal and transcriptional modifications are not explained. Thus, the detailed mechanism will be illustrated here (move the slider): \n \n Step 1. The 5’ cap is synthesized to protect RNA from exonucleolytic attack as well as to facilitate transcription and translation. Remaining part of the precursor RNA is transcribed from the viral genome in 5’ to 3’ direction.',
      
      'image': Image31,
      'afterImage': Image32
    },

{
      
      'description': 'The precursor RNA contains exons (orange) and introns (light blue).',
      'image': Image52
    },

{
    
      'description': 'Step 2. introns are removed/ not explicitly shown in TM',
    
      'image': Image33,
      'afterImage': Image34,
      'popup': {
        'title': 'Splicing visual representation',
        'content': [
          {
            'description': 'Other visual methods indicated the removed introns include dotted line, and interruptions that are connected by a “V” shape.',
            'image': Image35
          }
          
        ]
      }
    },

{
    
      'description': 'Step 3. It’s important to note that in vivo, the exons will be joined back together, and produce a shorter RNA transcript with a poly-A tail. This process is also not visually shown in most of the TMs.',
      
      'image': Image36,
      'afterImage': Image37
    },

{
      
      'description': 'Therefore, you may find out the RNAs in TM are usually illustrated as the intermediate step of RNA processing as shown in the purple box.',
      'image': Image38
    },

{
      
      'description': 'Why only showing the intermediate step of RNA processing in TM? \n \n The major advantage of this step is to help audiences to locate the exact positions of exons and removed introns to the viral genome. In this figure, you can use the guide lines to determine the length and position of every exons as well as introns.',
      'image': Image39
    },

{
      
      'description': 'Now, let’s refresh your memory and improve your understanding: please check all 5’ ends in this figure.',
      'image': Image40
    },

{
      
      'description': 'A common misconception would be defining the 5 prime end of RNA transcripts in TM. If you see the 5 prime end of RNA without a vertical bar or any visual cue, do not be confused, this is not the actual 5 prime end or transcription initiation site, instead, the interruption between RNA transcript, corresponding to the removed intron.',
      'image': Image41
    },

{
      'title': 'Using a hypothetical TM to understanding the Basic Visual Language - Determine RNA/DNA directionalities',
      'description': 'We already know the directionality of an RNA transcript, but what about the directionality of each DNA strand of the viral genome?',
      'image': Image42
    },

{
    
      'description': 'You can determine the directions of DNA double strands by following these two steps processes. \n Steps: \n 1. Look at your RNA and find its 3’ end (arrow head) and 5’ end (arrow tail). \n 2. Find the DNA single strand that adjacent to this RNA strand and this DNA strand reads in the opposite direction. Its complementary DNA strand has the same direction as the RNA strand. \n click the pop-up window to explore the details of molecular mechanism',
    
      'image': Image43,
      'popup': {
        'title': 'Viral genome directionalities',
        'content': [
          {
            'description': 'These two DNA strands have different names during the transcription. In this case, the top strand is referred to the coding strand (anti-sense), which serves as a template during transcription. The bottom strand is called the complementary strand (sense), which sequence is same as the RNA sequence (although all the Ts in RNA are replaced by Us).',
            'image': Image44
          },
           {
            'description': 'Coding strand is the only strand used in transcription since RNA polymerase can’t use both strands at each round of gene expression.',
            'image': Image45
          },
           {
            'description': 'Step1: When the transcription begins, the RNA polymerase will bind to DNA double helix.',
            'image': Image46
          },
           {
            'description': 'Step2: RNA polymerase will then unwind part of the DNA to find the coding strand.',
            'image': Image47
          },
           {
            'description': 'Step3: RNA polymerase will separate part of dsDNA to form a bubble like structure indicated as strands separation. let’s say you want to transcribe the gene from the bubble part of the DNA.',
            'image': Image48
          },
           {
            'description': 'Step4: RNA polymerase will scan through the coding DNA strand (antisense DNA) from 3’ to 5’, and RNA synthesis is from 5’ to 3’.',
            'image': Image49
          },
           {
            'description': 'Step5: Since newly synthesized RNA is complementary to the coding strand, scientists can hybridize the RNAs with their coding DNA strands to deduce the genome positions of these RNAs  – this experiment is also known as DNA-RNA hybridization. This is why the TM was designed in a way that all RNA transcripts are adjacent to their coding strands.',
            'image': Image50
          },
          
        ]
      }
    },

{
      
      'description': 'The directionality and the name of the DNA strands are relative to the directionality of the RNA transcript since the transcription can occurs at the both DNA strands, but in opposite direction. Please check the figure shown on your left side, The complementary strand for the top RNA transcript would be the coding strand for the bottom RNA transcript.',
      'image': Image51
    },



//module 2
{
      
      'image': Image63
    },

{
      'title': 'Module 2 Sections Menu:',
      'description': '1. Variations at 5’ end of the transcript \n 2. Variations at 3’ end of the transcript \n 3. Variations at both 5’ end and 3’ end of the transcript \n 4. Variations in body of the transcript \n 5. Variations in color and labels \n 6. Variations in translation regions \n \n 7. Visual Summary \n 8. Quiz- module 2',
      
      'image': Image62
    },

    {
      'title': 'Variations at 5’ end of the transcript - Alternative transcription promoter sites:',
      'description': 'Please review the first module and identify the meaning of this vertical bar shown in the blue box. \n\n Answer: This is a promoter site as well as represent the 5’ cap.',
      'image': Image64
    },
{
      'title': 'Variations at 5’ end of the transcript - Alternative transcription promoter sites',
      'description': 'However, in some viruses TM, you may have multiple transcription starting sites (TSS) as the figure shown - i.e. the multiple vertical bars. These alternative TSS all serve as potential promoters for polymerase to bind and to initiate transcription.',
      'image': Image65
    },

    {
      'title': 'Variations at 5’ end of the transcript - Alternative transcription promoter sites',
      'description': 'For each round of transcription, the polymerase can choose any one of these vertical bars as the transcription initiation site.',
      'image': Image66
    },
    {
      'title': 'Variations at 3’ end of the transcript - Alternative poly-A addition sites',
      'description': 'As we know so far, all the hypothetical RNAs introduced in module 1 are associated with only a particular viral genome region as the figure shown (in this case, the region from A to B).',
      'image': Image67
    },
 {
      'title': 'Variations at 3’ end of the transcript - Alternative poly-A addition sites',
      'description': 'Often in real TM, you can have the RNAs transcribed from the common promoter A, but end up with different 3’ ends. For instance, we have four RNAs with different poly-A sites B, C, D, E shown below. The RNA 3 with the poly-A site C contains the same RNA region as RNA 1 from A to B as well as associated with an additional viral genome region ranging from B to C.',
      'image': Image68
    },

{
      'title': 'Variations at 3’ end of the transcript - Alternative poly-A addition sites',
      'description': 'You might think these RNAs are alternatively and separately transcribed from viral genome in different lengths.',
      'image': Image69
    },

{
      'title': 'Variations at 3’ end of the transcript - Alternative poly-A addition sites',
      'description': 'This is not true, (recall the steps of poly-A addition shown in module 1). Instead, a long common precursor RNA including all these 4 poly-A addition sites is first transcribed from the viral genome. Then, many of these same precursor long RNA copies will be alternatively processed into either of these four shorter RNA transcripts. \n\n Click the pop-up window to check how this common precursor RNA get processed.',
    
      'image': Image70,
      'popup': {
        'title': 'Alternative Poly-A addition',
        'content': [
          {
            'description': 'Firstly, a long precursor RNA that contains all four poly-A sites is transcribed from a specific region of the viral genome',
            'image': Image71
          },
           {
            'description': 'Secondly, the polymerase can make several copies of this common precursor RNAs. There are four endonucleolytic cleavage sites in this precursor RNA and they are indicated by the black arrows. The precursor RNA can be cut at either of this four sites to produce a shorter RNA shown in four scenarios. Since the poly-A addition and the transcription almost occur simultaneously, the last cleavage site E also terminates the transcription of this precursor RNA.',
            'image': Image72
          },
          {
            'description': 'Lastly, the cleaved portions (i.e. the solid grey lines) will be degraded.',
            'image': Image73
          },
          {
            'description': 'This whole process is called alternative polyadenylation - a phenomenon that the mature RNA molecules with different 3′ ends originate from distinct polyadenylation sites of a single gene product/precursor RNA.',
            'image': Image74
          },
        ]
      }
    },

    {
      'title': 'Variations at both 5’ end and 3’ end of the transcript - Non-capped nor poly-A tail RNA transcript',
      'description': 'There is a group of a particular RNA transcript that shown with no arrow-tail nor arrowhead. This visual representation was not due to a mistake when they were designing the figure. The designer used different visual features based on the absence/presence of the transcriptional modifications to distinguish this type of RNA from other viral RNAs. The line indicated in orange is usually a non-capped and non-poly-A RNA since it is not translated during the viral life cycle.',
      'image': Image75
    },

{
      'title': 'Variations at both 5’ end and 3’ end of the transcript - Non-capped nor poly-A tail RNA transcript',
      'description': 'In adenovirus, this type of RNA is called viral associate RNA (VA RNA), which is transcribed by RNA polymerase class III. Although it lacks transcriptional modifications, VA RNA still has transcriptional initiation site, termination site & basic 5’ and 3’ directionality. (I highlighted the directionality in this figure for clarity.)',
      'image': Image76
    },
{
      'title': 'Variations in body of the transcript - Introduction to the mechanism of intron removal: RNA splicing',
      'description': 'Please review the first module and identify the meaning of this interruption shown in the blue box.\n\n Answer: This is a removed intron.',
      'image': Image77
    },
{
      'title': 'Variations in body of the transcript - Introduction to the mechanism of intron removal: RNA splicing',
      'description': 'Now, I add the intron back and I want you to understand that the intron removal is facilitated by a biological mechanism called RNA splicing. As the figure shown, initially you have a precursor RNA which includes exons, introns, and two different splice sites. I will use another type of visual representation to show the original biological structure clearly. These two splice sites will undergo chemical reactions to catalyze the intron removal.',
      'image': Image78
    },
{
      'title': 'Variations in body of the transcript - Introduction to the mechanism of intron removal: RNA splicing',
      'description': 'Although intron removal can occur automatically through chemical reactions, the splicing reactions need to perform in really high accuracy with the help of a splicing factory called spliceosome. Spliceosome guarantees that the coding information is not lost or altered due to the inaccurate splicing. The first step of RNA splicing is characterized by the assembling of the spliceosome, which controls the chemical reactions of the splicing. Then, we can observe an intron lariat - “loop structure.”',
      'image': Image79
    },
{
      'title': 'Variations in body of the transcript - Introduction to the mechanism of intron removal: RNA splicing',
      'description': 'The first chemical reaction detaches the 5’ end of the intron from splice donor site.',
      'image': Image80
    },
    {
      'title': 'Variations in body of the transcript - Introduction to the mechanism of intron removal: RNA splicing',
      'description': 'After the second reaction, the intron lariat will be released. At the same time, the exons are joining back together.',
      'image': Image81
    },
    {
      'title': 'Variations in body of the transcript - Introduction to the mechanism of intron removal: RNA splicing',
      'description': 'Lastly, you have your spliced RNA transcript.',
      'image': Image82
    },

    {
      'title': 'Variations in body of the transcript - Introduction to the mechanism of intron removal: RNA splicing',
      'description': 'Now, I assume you understand the mechanism of RNA splicing. In the figure, please identify all splice donor sites. \n\n Answer: Good! all light blue boxes you have chosen are the splice donor sites for this RNA transcript.',
      'image': Image83
    },
{
      'title': 'Variations in body of the transcript - Different splicing patterns ',
      'description': 'In real life TM, there are many other arrows stalk on top of each other and produce this weird looking representation. What does it mean?',
      'image': Image84
    },
{
      'title': 'Variations in body of the transcript - Different splicing patterns ',
      'description': 'Recall from previous section in this module, we already know they have alternative poly-A addition sites. Furthermore, these RNAs always share a common leader sequnce ',
      'image': Image85
    },
    {
      'title': 'Variations in body of the transcript - Different splicing patterns ',
      'description': 'To make this clearer, I added the common leader sequence back to each of the mRNA. You can see these mRNAs also have the different intron removal (splicing) patterns.',
      'image': Image86
    },
{
      'title': 'Variations in body of the transcript - Different splicing patterns ',
      'description': 'the introns are also added back to clearly show this unique pattern - also known as alternative intron removal or alternative splicing.',
      'image': Image87
    },



{
      'title': 'Variations in body of the transcript - Different splicing patterns',
      'description': 'Let me show you a good real world example: the adenovirus TM. You may realize that there are many different types of the transcript generated from a single region of the genome, and these variations of RNA transcripts is produced through the alternative intron removal mechanism (RNA splicing) as well as the alternative poly-A addition discussed in the previous section.  \n\n Click the button to open up a pop-up window describing the different splicing strategies.',
    
      'image': Image88,
      'popup': {
        'title': 'Splicing patterns',
        'content': [
          {
            'description': 'RNA Splicing can be divided into two large categories: constitutive splicing (red box) or alternative splicing (purple box).',
            'image': Image89
          },
           {
            'description': 'Constitutive splicing is the classical method of RNA splicing and itself is not sufficient for the RNA diversity.',
            'image': Image90
          },
          {
            'description': 'In order to increase the diversity of transcriptome, alternative splicing is used to produced different protein products with same encoded gene. (I only introduce three common alternative splicing examples.) \n \n Exon skipping: bypass the exon 2, and join the exon 1 and 3 together. \n\n Alternative donor site splicing: An alternative 5 prime splice junction (donor site) is used, changing the 3 prime boundary of the upstream exon. \n\n Alternative acceptor site splicing: An alternative 3 prime splice junction (acceptor site) is used, changing the 5 prime boundary of the downstream exon.',
            'image': Image91
          },
          {
            'description': 'An interesting fact for alternative splicing is that we define introns and exons differently in each mRNA. Please check the RNA transcript region shown between these two dotted guiding lines. You can see this part of the fourth mRNA’s exon belongs to the intron region of mRNA 1, 2, 3. Therefore, a misconception from the textbook illustration would be illustrating the exons and introns in fixed positions. In reality, The lengths and regions of exon and intron are vary and they are defined according to the different splicing patterns.',
            'image': Image92
          },
        ]
      }
    },

    {
      'title': 'Variations in body of the transcript - Different splicing patterns ',
      'description': 'You may ask why they get rid of the common leader sequence for RNA transcripts shown in the red box and make the diagram so confusing?',
      'image': Image93
    },
 {
    
      'image': Image94
    },

     {
    
      'image': Image95
    },

    {
      'title': 'Variations in body of the transcript - Thickness of transcript ',
      'description': 'Different transcripts can have difference thickness of their arrow body. The thickness of arrow represents the abundance of the transcripts get produced. As you can see, the RNA with a thicker arrow will be transcribed in a high level during the viral life cycle as opposed to the RNA with a thinner arrow.',
      'image': Image96
    },

 {
      'title': 'Variations of RNAs- color',
      'description': 'Some viruses do not express all the genes at once. Instead they tightly control their gene expression at different stages of infection. TMs visually communicate these stages using techniques such as color coding with labels or labels only. In this example, we are using different colors as well as labels to indicate three transcription stages',
      'image': Image97
    },

{
      'title': 'Variations of RNAs- text label ',
      'description': 'Adjacent to each RNA transcript, you will find a label indicating the identity of the corresponding protein product. For instance, RNA transcript with “hexon” label beside it indicating that this RNA will finally translate into the hexon protein used to build the viral capsid.',
      'image': Image98
    },




    
{
      'title': 'Variations in translation regions ',
      'description': 'In a TM, you may see ORF or CS labeled beside the transcript. These are the language used to describe the translation regions. ORF means open reading frame, which consists of a DNA sequence starting with a start codon and ending with a stop codon. Usually, ORF refers to a DNA/RNA sequence that has potential ability to be translated after RNA processing. Therefore, not all regions of ORF get translated since we might have introns in the ORF of unprocessed RNA molecule.',
      'image': Image99
    },
{
      'title': 'Variations in translation regions ',
      'description': 'On the contrary, CS, the coding sequence, is the actual coding region that can be translated into the protein. The diagram shown below helps to distinguish these two terms.',
      'image': Image100
    },

{
      'title': 'Variations in translation regions ',
      'description': 'In addition, different ORFs will be read in different scenarios to produce different RNAs.\n\n In this case, both ORF1 and ORF 2 has ability to be translated depends on different splicing mechanisms. If the stop codon of ORF 1 within intron is spliced out from the transcript, the stop codon of ORF 2 will be used during this mRNA translation. Therefore, ORF 2 will be read in this scenario. All the exons in this ORF 2 will be translated into protein.',
      'image': Image101
    },
{
      'title': 'Variations in translation regions ',
      'description': 'In different case, if the stop codon of ORF 1 within intron is retained during alternative splicing, the stop codon of ORF 1 will be used during this mRNA translation. Therefore, ORF 1 will be read in this scenario. Although all exons are still transcribed, the exons regions located after the ORF 1 will not be translated into protein.',
      'image': Image102
    },
{
      'title': 'Variations in translation regions ',
      'description': 'You may also realize that some regions of the exons are not going to be translated because these regions are located beyond the ORF. These regions have a specific name: untranslated region (UTR). \n\n If UTR locates at 5 prime end, we call it 5 prime UTR. Same with the 3 prime end.',
      'image': Image103
    },

  ],
  focus: 0,
  currentPage: 0,
  sliding: false
};

function rootReducer(state = initialState, action) {
  if (action.type === CHANGE_FOCUS) {
    if (action.focus >= 0 && action.focus < state.pageData.length) {
      state = Object.assign({}, state, {
          focus: action.focus
      });
    }
  }
  else if (action.type === CHANGE_PAGE) {
    if (action.currentPage >= 0 && 
        action.currentPage < state.pageData.length) {
      state = Object.assign({}, state, {
          currentPage: action.currentPage,
          sliding: true
      });
    }
  }
  else if (action.type === SLIDING_END) {
    clearInterval(action.intervalID);
    state = Object.assign({}, state, {
        sliding: false
    });
  }

  return state;
}

export default rootReducer;